File: solutions/1005/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sun Nov  7 11:23:50 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <numeric>
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
private:
    static bool cmp(int a, int b) {
        return abs(a) > abs(b);
    }
public:
    int largestSumAfterKNegations(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end(), cmp);
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] < 0 && k > 0) {
                nums[i] *= -1;
                k--;
            }
        }
        if (k % 2 == 1)
            nums[nums.size()-1] *= -1;
        int result = 0;
        result = accumulate(nums.begin(), nums.end(), result);
        return result;
    }
};

TEST(Test1005, SimpleTest) {
    vector<int> nums{
        4,2,3
    };
    const int k = 1;
    Solution s; 
    EXPECT_EQ(s.largestSumAfterKNegations(nums, k), 5);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/1005/main.py
from typing import List 

class Solution:
    def largestSumAfterKNegations(self, nums: List[int], k: int) -> int:
        nums = sorted(nums, key=abs, reverse=True)
        for i in range(len(nums)):
            if nums[i] < 0 and k > 0:
                nums[i] *= -1
                k -= 1
        if k % 2:
            nums[-1] *= -1
        result = 0
        for i in range(len(nums)):
            result += nums[i]
        return result

if __name__ == "__main__":
    nums = [4,2,3]
    k = 1 
    mySol = Solution()
    assert mySol.largestSumAfterKNegations(nums, k) == 5
